src/cargo/core/package.rs

```diff
@@ -571,6 +558,66 @@ impl<'cfg> PackageSet<'cfg> {
         Ok(())
     }
 
+    /// Check if there are any dependency packages that do not have any libs.
+    pub(crate) fn no_lib_pkgs(
+        &self,
+        resolve: &Resolve,
+        root_ids: &[PackageId],
+        has_dev_units: HasDevUnits,
+        requested_kinds: &[CompileKind],
+        target_data: &RustcTargetData<'_>,
+        force_all_targets: ForceAllTargets,
+    ) -> CargoResult<Vec<&Package>> {
+        let mut ret = vec![];
+
+        root_ids.iter().for_each(|pkg_id| {
+            PackageSet::filter_deps(
+                *pkg_id,
+                resolve,
+                has_dev_units,
+                requested_kinds,
+                target_data,
+                force_all_targets,
+            )
+            .for_each(|(package_id, _)| {
+                if let Ok(dep_pkg) = self.get_one(package_id) {
+                    if !dep_pkg.targets().iter().any(|t| t.is_lib()) {
+                        ret.push(dep_pkg);
+                    }
+                }
+            });
+        });
+
+        Ok(ret)
```

Comment:
Would it be useful to have the error message contain the relevant root_id package for which the dep is inside? What do the root_ids represent here? From a quick scan, it seems like it's for workspaces with multiple members?

Perhaps error message could read
```
[WARNING] `foo` has invalid dependency `bar`. `bar` has no lib package.
```
Can't quite figure out the wording, but something including `foo` here (using your testcase as the example) would help the user in workspaces with several members.

Reply:
If we use the root_id warning here it may get long. If we need to make this change, then we should probably have a separate warning for each root pkg. It may need to be done in conjunction with your first comment.

Reply:
Maybe I'll address it tonight.

Reply:
Also, if a root_id package has multiple packages without libs, how do we need to display it?

cc: @ehuss Do you have any suggestions for the display of warnings here?

Reply:
One proposal:
In this example - the workspace has 2 member crates `foo1` and foo2`. `bar{n}` are all invalid dependencies
```
[WARNING] `foo1` has invalid dependency `bar1`. `bar1` has no lib package.
[WARNING] `foo1` has invalid dependency `bar2`. `bar2` has no lib package.
[WARNING] `foo2` has invalid dependency `bar1`. `bar1` has no lib package.
[WARNING] `foo2` has invalid dependency `bar3`. `bar3` has no lib package.
```

Where foo1's Cargo.toml
```
[dependencies]
bar1 = *
bar2 = *
```

foo2's Cargo.toml
```
[dependencies]
bar1 = *
bar3 = *
```

Reply:
Done! Thank you for your help!